content/2020-05-22-book-review-writing-an-interpreter-in-go/index.md

```diff
@@ -0,0 +1,581 @@
+    Token: p.curToken,
+    Value: p.curTokenIs(token.TRUE)
+    }
+}
+```
+
+I wrote a question mark next to the `token.TRUE` because I was confused on why this works even when parsing a `false` boolean. Again, this is probably do to my lack of understanding here and another area I would focus on again.
+
+#### grouped expressions
+
+Some how, we can group expressions like `(5 + 5) * 2` to influence the parsing to our liking. We add a function called `parseGroupedExpression` and it "just works." I couldn't tell you why though. Here's the function in case you can figure it out:
+
+```go
+func (p *Parser) parseGroupedExpression() ast.Expression {
+  p.nextToken()
+  exp := p.parseExpression(LOWEST) // this LOWEST thing I don't get
+
+  if !p.expectPeek(token.RPAREN) {
+    return nil
+  }
+
+  return exp
+}
+```
+
+#### block statements
+
+These are a series of statements which are surrounded by `{` and `}`.
+
+#### infix parse functions
+
+I wrote a note later where he says,
+
+> Yes, we need to register an `infixParseFn` for `token.LPAREN`. This way we parse the expression that is the function (either an identifier, or a function literal), then check for an `infixParseFn` associated with `token.LPAREN` and call it with the already parsed expression as argument. And in this `infixParseFn` we can then parse the argument list.
+>
+> -Ball 122
+
+I'm blocking myself here. In my mind, I think "infix === things like +, -, \*, /" so it feels strange here to use an LPAREN or "(" as an infix, but I guess infix can also mean in between two other characters. Example: `add(5, 5)` where the `(` is in between the "d" and the "5". Still, future me should revisit this.
+
+#### serialization
+
+Thorsten mentions this,
+
+> What we want is an AST that (serialized as a string)...
+>
+> -Ball 89
+
+And I thought, "Do I really know what this means?" And the answer is no, no I do not. Wikipedia [explains it](https://en.wikipedia.org/wiki/Serialization) well. I'll summarize it as taking something, translating it into a data structure or object and storing it for later use (possibly restructuring it).
+
+### Chapter 3 - Evaluation
+
+This is the part where the abstract syntax trees are turned into something meaningful.
+
+#### interpreters vs compilers
+
+> The notion of an interpreter as something that doesn't leave executable artifacts behind (in contract to a compiler, which does just that) gets fuzzy real fast when looking at the implementations of real-world and highly-optimized programming languages.
+>
+> -Ball 132
+
+I found this super helpful (at least comparing it to say Rust). Interpreter has one-step. Compilers have two.
+
+#### tree-walking interpreters
+
+I wrote down this because I wanted to look up a few examples of programming languages that fall into this category. I don't know how trustworthy [these Reddit comments](https://www.reddit.com/r/ProgrammingLanguages/comments/dfi9lk/viability_of_treewalk_interpreters/) are but here are a few notes:
```

Comment:
```suggestion
I wrote this down because I wanted to look up a few examples of programming languages that fall into this category. I don't know how trustworthy [these Reddit comments](https://www.reddit.com/r/ProgrammingLanguages/comments/dfi9lk/viability_of_treewalk_interpreters/) are but here are a few notes:
```